plugins/catalog-backend/src/service/createRouter.ts

```diff
@@ -135,24 +138,92 @@ export async function createRouter(
-          url.searchParams.delete('offset');
-          url.searchParams.set('after', pageInfo.endCursor);
-          res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+        // When pagination parameters are passed in, use the legacy slow path
+        // that loads all entities into memory
+
+        if (pagination) {
+          const { entities, pageInfo } = await entitiesCatalog.entities({
+            filter,
+            fields,
+            order,
+            pagination,
+            credentials,
+          });
+
+          // Add a Link header to the next page
+          if (pageInfo.hasNextPage) {
+            const url = new URL(`http://ignored${req.url}`);
+            url.searchParams.delete('offset');
+            url.searchParams.set('after', pageInfo.endCursor);
+            res.setHeader('link', `<${url.pathname}${url.search}>; rel="next"`);
+          }
+
+          res.json(entities);
+          return;
         }
 
-        // TODO(freben): encode the pageInfo in the response
-        res.json(entities);
+        // For other read-the-entire-world cases, use queryEntities and stream
+        // out results.
+
+        // The write lock is used for back pressure, preventing slow readers
+        // from forcing our read loop to pile up response data in userspace
+        // buffers faster than the kernel buffer is emptied.
+        // https://nodejs.org/api/http.html#http_response_write_chunk_encoding_callback
+        const locks: { writeLock?: DeferredPromise } = {};
+        const controller = new AbortController();
+        const signal = controller.signal;
+        req.on('end', () => {
+          controller.abort(new Error('Client closed connection'));
+          locks.writeLock?.resolve();
+          delete locks.writeLock;
+        });
+
+        const responseStream = createEntityArrayJsonStream(res);
+        const limit = 10000;
+        let cursor: Cursor | undefined;
+
+        try {
+          do {
+            const result = await entitiesCatalog.queryEntities(
+              !cursor
+                ? { credentials, fields, limit, filter, orderFields: order }
+                : { credentials, fields, limit, cursor },
+            );
+
+            if (result.items.length) {
+              await locks?.writeLock;
+
+              signal.throwIfAborted();
+
+              expandLegacyCompoundRelationRefsInResponse(result.items);
+              if (!responseStream.send(result.items)) {
+                // The kernel buffer is full. Create the lock but do not await it
+                // yet - we can better spend our time going to the next round of
+                // the loop and read from the database while we wait for it to
+                // drain.
+                locks.writeLock = createDeferred();
+                res.once('drain', () => {
+                  locks.writeLock?.resolve();
+                  delete locks.writeLock;
+                });
```

Comment:
what about doing the following so that we won't need any other locks spread in the code?

```suggestion
                const drainLock = createDeferred();
                res.once('drain', () => drainLock.resolve());
                await drainLock;
```

of course we won't read the next iteration of entities, but the code will be much simpler.

Reply:
Yeah exactly

Reply:
we could do a whole dance where we both subscribe to end and drain, and unsubscribe from end, etc - but it's gonna be a bit of a bigger piece of code then and it'll aggressively subscribe and unsubscribe at a high rate, not sure if worth it